jgrapht-core/src/main/java/org/jgrapht/alg/CyclicTransitiveReduction.java

```diff
@@ -0,0 +1,297 @@
+ * also works correctly for directed graphs containing cycles because the simple {@link TransitiveReduction} algorithm
+ * fails in this case, falsely deleting all edges from a cyclical graph.
+ * <p></p>
+ * This algorithm utilises existing JGraphT functionality such as:
+ * <ul>
+ *   <li>transitive reduction for DAGs (directed acyclical graphs)</li>
+ *   <li>cycle detection</li>
+ *   <li>strong connectivity inspector incl. graph condensation</li>
+ * </ul>
+ * In addition to that the algorithm performs edge pruning both within and between SSCs (strongly connected
+ * components) as well as condensed graph expansion (basically the inverse operation) back into the original graph.
+ * <p></p>
+ * The result will be a minimal, transitively reduced graph, only containing edges which were also present in the
+ * original graph. This comes at a cost of performance because it requires additional checks which could be omitted if
+ * we would relax the requirement to just return a minimal graph with identical reachability. Then for example it would
+ * be permitted
+ * <ul>
+ *   <li>
+ *     to replace any SCC by a simple cycle connecting all vertices, potentially containing synthetical edges. Instead,
+ *     this algorithm guarantees to find an existing Hamiltonian cycle within each SCC and reduces the SCC to it.
+ *   </li>
+ *   <li>
+ *     to replace redundant edges between SCCs by a single, potentially synthetical edge between any two source and
+ *     target vertices. Instead, this algorithm guarantees that it retains an one existing link between SCCs, removing
+ *     all other edges as redundant.
+ *   </li>
+ * </ul>
+ * <p></p>
+ * <b>Implementation note:</b> This algorithm was compared with an improved version of an existing algorithm found in a
+ * C library from the <a href="https://www.win.tue.nl/emcmc/cutter/">CUTTER project</a> related to the paper
+ * <a href="https://bmcbioinformatics.biomedcentral.com/track/pdf/10.1186/1471-2105-13-281.pdf">Efficient reconstruction
+ * of biological networks via transitive reduction on general purpose graphics processors</a> and ported to Java,
+ * operating directly on int or boolean arrays. The performance was comparable only for small cyclical digraphs, but the
+ * (non-parellel) CUTTER algorithm, when compared to this one, performs worse and worse for bigger networks with more
+ * vertices and more SCCs. Furthermore, for DAGs CUTTER loses even more dramatically because it uses the Warshall
+ * transitive closure algorithm which is very slow for large DAGs. So there was no apparent advantage in keeping the
+ * array-based algorithm despite my initial hope that it might be faster than this algorithm. Hence, I decided to keep
+ * and improve this one instead.
+ *
+ * @param <V> vertex type
+ * @param <E> edge type
+ * @author Alexander Kriegisch
+ */
+public class CyclicTransitiveReduction<V, E> {
+  protected final Graph<V, E> directedGraph;
```

Comment:
I imagine this class to be specialised by improved implementations, hence `protected`. It would also allow the methods to be directly unit-tested, which is of course not a compelling reason, only a positive side-effect. I am really struggling to understand some of the comments: one package-scoped method is not open enough, these ones too open. Is this level of code review really necessary before I receive any comment about the core functionality as such? This kind of review process is disheartening and exactly what I was afraid of. Like I said in the very beginning: Why not fix the code functionally if there are any real bugs, then accept the code and improve it yourself afterwards if you feel it does not comply to some internal rules, instead of pulling my strings like a puppet's and using more words and time than the actual code change would take?

Reply:
Sorry to disappoint--that's not how we work, and never will be.  The first code review provides a good sense on both sides for whether it will be a productive relationship going forward, and if either side doesn't think it will be worth the effort for future contributions, then that becomes clear pretty fast.

Reply:
If you cannot see the value of my contribution as it is, reject it. My life does not depend on you accepting it and this ought not be an assessment centre. I have no plans to become a regular contributor here. I merely used the library for the first time, notices some missing features, implemented them in my own classes first and then reconsidered, trying to return something to the OSS community. I tried to fit it into the library to the best of my knowledge, added Javadoc and even a test suite. I could have stopped at my first piece of sample code in #1008 which already did what I wanted. My personal use case are very small graphs, everything I did in order to make this better, faster etc. was purely for the benefit of this project. My benefit is zero.

I acknowledge the benefit of code reviews. But there are such and such code reviews. I received one comment regarding functionality, the one about edge set vs vertex set and improving performance. That one was valuable. The rest consists of minor concerns, none of which should stop the contribution from being accepted. Even so, I accommodated several of them. But I do not see myself as the reviewer's slave. We are peers, this process should be on eye level. There is a limit to my willingness to comply, and we have exceeded it by far now.

I am still open to further comments and which consider each one individually without holding a grudge from the ones I disagree with. If you spot a bug or another significant deficiency in my contribution and it is in my power to fix it myself, I shall do so.

Reply:
Truly, we are forever in your debt, and I heave a heartfelt sigh of relief to hear that you have no plans to become a regular contributor here.

Reply:
Also, as I mentioned before, we generally don't reject incomplete contributions; we just leave them pending, and if someone else comes along and finishes them, then we merge them.  There's no hurry--we've been around for more than 15 years, and we'll probably be around 15 years from now, and graph theory is not going away.

Reply:
I don't wish anyone to feel **indebted** to me, no matter if in a sarcastically meant or true way. What I wish for is to feel **welcome** as a contributor, to get substantial comments and get to collaboratively work on code. Remember, I even offered to give commit rights to my fork to maintainers here right at the beginning of this PR process, so they can collaborate in making the PR better by themselves instead of writing dozens of comments about minor details, showing off their position of power and a gatekeeper mentality which scares off anyone considering contributing to this project.

> (...) we generally don't reject incomplete contributions; we just leave them pending (...) There's no hurry--we've been around for more than 15 years, and we'll probably be around 15 years from now (...)

You are just being condescending. Like I said, gatekeeper mentality: "Do exactly as I say or your contribution shall stay in limbo forever or until you yield."

Reply:
We welcome all contributors who demonstrate an interest in working with our established process.  Others are welcome to start their own projects, contribute to other projects, etc.  There's no point trying to make a square peg fit a round hole.

Reply:
I was actively participating in your established process, adapting my code where I agreed it was helpful to mature the code and make it more acceptable to the project. But there is a limit to that. I think that established process is meant to stop bad code from being accepted into the code base and I applaud and support that. What you are doing is to use that process in order to show off your position of power, treat fellow developers like little kids. Who says that a square peg is a bad thing? Why does the hole have to be round? Why don't you make it round if you feel it is too square and I invite you to do so by yourself? You have yet to name a single functional issue in my contribution. My code is clean and maintainable enough. Go on obsessing about access modifiers, exception types and transitional package names, but please do so without me.

_The most dangerous phrase in the language is, "We've always done it this way."_
Grace Hopper